### PR TITLE
First attempt at creating queries for LSC tasks

### PR DESCRIPTION
**Pull request by benzvera**
Created a task pool for lifelog queries and implemented the first one. Tags, text-co-embeddings and boolean searches can be implemented as objects now.

### CODE REVIEW

File: src/main/kotlin/com/github/silvanheller/tempAlgEval/main/SpecsBuilder.kt [hunk 1]
```diff
@@ -132,63 +88,111 @@ class SpecsBuilder {
...
+                        Pool.time
+                    )
+                )
+            }
+            else -> {
+                tasks.add(
+                    SpecificationTask(
+                        "",
+                        TaskType.KIST,
+                            listOf(Pool.result) as List<Result>,
```
Comment (silvanheller) [inline src/main/kotlin/com/github/silvanheller/tempAlgEval/main/SpecsBuilder.kt:187]: I think this is a consequence of declaring the `result` in `Pool` as `Any`.  I would simply declare it as `List<Result>` (which is then just length 1 for VBS-tasks), which leads to clearer code and you don't have to cast to List<> here

File: src/main/kotlin/com/github/silvanheller/tempAlgEval/main/TaskPoolLSC.kt [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+ * This file contains a pool of queries for solving tasks of the lsc in 2019.
+ */
+
+fun listOfPoolsLSC() =
+        listOf(
+                PoolWithTime(
+                        "LSC26",
+                        listOf(Result(
+                                "is_2016-08-20",    // ObjectId = one day
+                                0F,                 // start and end are always 0
```
Comment (silvanheller) [inline src/main/kotlin/com/github/silvanheller/tempAlgEval/main/TaskPoolLSC.kt:16]: This needs to specify start and end, as in LSC tasks we're looking for specific events within the day, and not just the whole day.

File: src/main/kotlin/com/github/silvanheller/tempAlgEval/main/TaskPoolLSC.kt [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+                                                                TermObject( //Tag
+                                                                        QueryCategories.text,
+                                                                        listOf(
+                                                                                QueryTypes.concept
+                                                                        ),
+                                                                        listOf(
+                                                                                Tag(
+                                                                                        "driving",
+                                                                                        "driving",
+                                                                                        "driving"
```
Comment (silvanheller) [inline src/main/kotlin/com/github/silvanheller/tempAlgEval/main/TaskPoolLSC.kt:33]: The Tag descriptions and name are not necessary for retrieval

File: src/main/kotlin/com/github/silvanheller/tempAlgEval/main/TaskPoolLSC.kt [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+                                                                                ),
+                                                                                Tag(
+                                                                                        "car_interior",
+                                                                                        "car_interior",
+                                                                                        "car_interior"
+                                                                                )
+                                                                        )
+                                                                ),
+                                                                TermObject(//Boolean search
+                                                                        QueryCategories.image,
```
Comment (silvanheller) [inline src/main/kotlin/com/github/silvanheller/tempAlgEval/main/TaskPoolLSC.kt:43]: image seems incorrect to me :) You probably want to add a new a category called `filter`, which is what was previously used for VBS interaction logging.

File: src/main/kotlin/com/github/silvanheller/tempAlgEval/model/enums.kt [hunk 5]
```diff
@@ -33,13 +34,14 @@ enum class QueryCategories {
...
-    concept,
-    jointEmbedding,
-    sketch,
-    semanticSegmentation,
+    metadata,               //Boolean
+    OCR,                    //Optical Character Recognition
+    ASR,                    //Automatic Speech Recognition
+    conceptBase64Legacy,    //Tags in base64
+    concept,                //Tags
+    jointEmbedding,         //
```
Comment (silvanheller) [inline src/main/kotlin/com/github/silvanheller/tempAlgEval/model/enums.kt:42]: This means text is mapped to an image / video representation space

File: src/main/kotlin/com/github/silvanheller/tempAlgEval/model/enums.kt [hunk 6]
```diff
@@ -52,15 +54,33 @@ enum class CineastCategories {
...
 
 enum class EvaluationMetrics {
-    VBS2019
+    VBS2019,
+    LSC2019
+}
+
+enum class Dataset {
+    VBS,
+    LSC
```
Comment (silvanheller) [inline src/main/kotlin/com/github/silvanheller/tempAlgEval/model/enums.kt:83]: This should probably specify the years (LSC2018 & 2019 used the same dataset, 2020 & 2021 used the same dataset, and 2022 will use a new dataset).

File: src/main/kotlin/com/github/silvanheller/tempAlgEval/model/models.kt [hunk 7]
```diff
@@ -139,14 +180,30 @@ data class QueryTerm(
     val tags: List<Tag> = listOf(),
     val sliderSetting: Float = 0f,
     val categories: List<CineastCategories>,
-    val data: String,
+    val data: Any
 ) : Model
 
+open class QueryTermType()
```
Comment (silvanheller) [inline src/main/kotlin/com/github/silvanheller/tempAlgEval/model/models.kt:186]: I don't fully understand the reasoning behind this refactoring, maybe we can discuss this offline

File: src/main/kotlin/com/github/silvanheller/tempAlgEval/query/websocket/QueryBuilder.kt [hunk 8]
```diff
@@ -91,73 +92,156 @@ class QueryBuilder {
...
+                                logger.info(
+                                    "Unknown category occurred $category"
+                                )
                                 continue
                             }
                         }
                     }
                     var tags: List<Tag> = listOf()
+                    var dataString = ""
                     if (cineastType == CineastQueryTermType.TAG) {
```
Comment (silvanheller) [inline src/main/kotlin/com/github/silvanheller/tempAlgEval/query/websocket/QueryBuilder.kt:167]: use `when` also for CineastTypes

File: src/main/kotlin/com/github/silvanheller/tempAlgEval/scoring/algorithms/vitrivr/VitrivrScoringAlgorithm.kt [hunk 9]
```diff
@@ -36,7 +24,7 @@ class VitrivrScoringAlgorithm(queryStorage: QueryStorage) : ScoringAlgorithm(que
         private val logger = LogManager.getLogger(Companion::class.java)
     }
 
-    override fun scoreInternal(): ResultScoreTuple {
+    override fun scoreInternal(dataset: Dataset): ResultScoreTuple {
```
Comment (silvanheller) [inline src/main/kotlin/com/github/silvanheller/tempAlgEval/scoring/algorithms/vitrivr/VitrivrScoringAlgorithm.kt:27]: I think an easier structure would be to have the `dataset` as a parameter for the `ScoringAlgorithm`, and then the `parseObjectId` method could take `dataset` as a parameter which you can access with `this.dataset`. This means you don't have to duplicate the code handling parsing for different datasets across the scoring algorithms.

**Review by silvanheller — CHANGES_REQUESTED** [review]
Looks promising, thanks for the work! I left some comments. One general feedback:
- The `checkstyle-idea.xml` and `google-java-format.xml` files should not be in the repo, as we're using ktlint

File: src/main/kotlin/com/github/silvanheller/tempAlgEval/main/TaskPoolLSC.kt (hunk 3 above)
Comment (benzvera) [reply src/main/kotlin/com/github/silvanheller/tempAlgEval/main/TaskPoolLSC.kt]: The tag name is important for transforming the tag into a text co-embedding

**Review by silvanheller — CHANGES_REQUESTED** [review]

**Pull request merged**